src/analyses/base.ml

```diff
@@ -803,13 +801,13 @@ struct
         match (eval_rv a gs st n) with
         | `Address adr ->
           (if AD.is_null adr
-           then M.warn_each ~must:true ~warning:(M.Warning.Behavior.Undefined.nullpointer_dereference ()) ()
+           then M.error_each ~category:M.Category.Behavior.Undefined.nullpointer_dereference ~tags:[M.Tag.CWE 476] "Must dereference NULL pointer"
```

Comment:
Maybe it makes sense to abstract warnings that have multiple calls in order to avoid inconsistencies.
A `nullpointer_dereference` warning should include its CWE and text.

Reply:
I'm not sure how necessary this would be because there shouldn't be duplication in warnings in the first place: each one should be emitted by an analysis responsible for it. And even here there are subtle differences between the two: the error is for a must-dereference (mentioned in the message) and the warning is for a may-dereference (mentioned in the message).

And coupling categories with CWEs etc is precisely what we want to avoid in some other places, for example integer overflows. The complete messaging for those is not yet there, but the idea is as follows. Unlike signed overflows, unsigned overflows are not undefined behavior, but they are still considered under the same CWE. While signed overflows should also have an undefined behavior category tag in addition to the integer overflow one. And to be extra precise, there's separate CWEs for overflow and underflow.

So all the subtle variations intentionally should have different tags (and probably messages).

Reply:
> And even here there are subtle differences between the two: the error is for a must-dereference (mentioned in the message) and the warning is for a may-dereference (mentioned in the message).

The text then duplicates the information given by warn/error.

> And coupling categories with CWEs etc is precisely what we want to avoid in some other places, for example integer overflows.

Makes sense, but there you can still have different calls.

> I'm not sure how necessary this would be because there shouldn't be duplication in warnings in the first place: each one should be emitted by an analysis responsible for it.

The two calls here are identical besides warn/error.
There may be several analysis reporting the same warning.
Maybe it's not that common, I just spotted this bit and imagined it may be a source of duplicated code.
But it's also not clear how to best abstract it. Could introduce warn/error functions that take pre-defined message records.

Reply:
> The text then duplicates the information given by warn/error.

This is true for must/may, but it's probably for the user's benefit, such they don't have to deduce it from whether it's an error or just a warning. The exact difference between errors and warnings might be analysis-dependent though and more general distinction than must/may. Attaching must/may to every message also doesn't make sense.

> There may be several analysis reporting the same warning.

The fact that different analyses report the same thing sounds like a duplication in the first place though.

> But it's also not clear how to best abstract it. Could introduce warn/error functions that take pre-defined message records.
> It may also be worth thinking about whether it's viable to define all warnings/errors (not debug) in one place (similar to `defaults.ml`) and then just reference them.

This would be a possible direction indeed if we want to very strictly enumerate all our warnings, a la some compilers or analyzers which specify unique IDs for each warning type. But given that most of the messages still have unknown category and haven't been even properly reviewed or categorized broadly, I think that's too much of a stretch for now.